### PR TITLE
wasmtime-wasi: replace FilePerms, DirPerms with single bit FsPerms

### LINKED ISSUE #14003: wasmtime-wasi: pair of DirPerms, FilePerms should be replaced with a single bit permitting mutation
wasmtime-wasi accepts a pair of what are effectively booleans, [DirPerms](https://docs.rs/wasmtime-wasi/latest/wasmtime_wasi/filesystem/struct.DirPerms.html) and [FilePerms](https://docs.rs/wasmtime-wasi/latest/wasmtime_wasi/filesystem/struct.FilePerms.html), to set permissions on a [`preopened_dir`](https://docs.rs/wasmtime-wasi/latest/wasmtime_wasi/struct.WasiCtxBuilder.html#method.preopened_dir). This permissions structure is specific to wasmtime-wasi's implementation, and is not specified by WASI. It is very poorly specified by its own documentation (a sum total of two doc strings on the DirPerms flags) and that documentation is inconsistent with the actual behavior - DirPerms::MUTATE without FilePerms::WRITE does not permit creating a new file.

The weird corner case created by DirPerms::MUTATE without FilePerms::WRITE has been the subject of multiple recent Wasmtime security advisories: [GHSA-2r75-cxrj-cmph](https://github.com/advisories/GHSA-2r75-cxrj-cmph) [GHSA-4ch3-9j33-3pmj](https://github.com/bytecodealliance/wasmtime/security/advisories/GHSA-4ch3-9j33-3pmj). I conducted an informal survey while remediating these, and AFAIK no production user of wasmtime-wasi is using that combination of permissions, nor any of the other "weird" cases (any of FilePerms or DirPerms being empty).

The set that actually seems to be useful to wasmtime-wasi's users is either that an entire preopen is is read-write, or is read-only. So, this issue proposes deleting the DirPerms and FilePerms structures, and replacing them with a single boolean that permits mutation of the entire preopen. This should hopefully make the permissions structure easy for both the implementation to enforce and for users to understand, and reduce the amount of time maintainers spend addressing security advisories for behaviors that nobody uses.

### PR DESCRIPTION
**Pull request by pchickey**
Closes #14003 

wasmtime-wasi accepted a pair of flag-sets, [DirPerms](https://docs.rs/wasmtime-wasi/latest/wasmtime_wasi/filesystem/struct.DirPerms.html) and [FilePerms](https://docs.rs/wasmtime-wasi/latest/wasmtime_wasi/filesystem/struct.FilePerms.html), to set permissions on a [preopened_dir](https://docs.rs/wasmtime-wasi/latest/wasmtime_wasi/struct.WasiCtxBuilder.html#method.preopened_dir). This permissions structure is specific to wasmtime-wasi's implementation, and is not specified by WASI. It is very poorly specified by its own documentation (a sum total of two doc strings on the DirPerms flags) and that documentation is inconsistent with the actual behavior - DirPerms::MUTATE without FilePerms::WRITE does not permit creating a new file.

The weird corner case created by DirPerms::MUTATE without FilePerms::WRITE has been the subject of multiple recent Wasmtime security advisories: [GHSA-2r75-cxrj-cmph](https://github.com/advisories/GHSA-2r75-cxrj-cmph) [GHSA-4ch3-9j33-3pmj](https://github.com/bytecodealliance/wasmtime/security/advisories/GHSA-4ch3-9j33-3pmj). I conducted an informal survey while remediating these, and AFAIK no production user of wasmtime-wasi is using that combination of permissions, nor write-without-read file permissions (covered by only one test), nor any of the other "weird" cases (any of FilePerms or DirPerms being empty).

The set that actually seems to be useful to wasmtime-wasi's users is either that an entire preopen is is read-write, or is read-only. So, this PR deletes DirPerms and FilePerms structures, and replaces them with a single boolean that permits mutation of the entire preopen, rendered as `enum FsPerms { ReadOnly, ReadWrite }`. This should hopefully make the permissions structure easy for both the implementation to enforce and for users to understand, and reduce the amount of time maintainers spend addressing security advisories for behaviors that nobody uses.

This PR was pretty mechanical:
* replacing all uses of the pair of FilePerms, DirPerms with a single FsPerms, starting in `WasiCtxBuilder::preopened_dir` and following that thread throughout the crate.
* any checks that required Read permissions and would return NotPermitted otherwise are erased - Read permission is always available
* any checks that checked for absence of FilePerms::WRITE or DirPerms::MUTATE now call `FsPerms::write_not_permitted(&self) -> bool`

Slightly non-obvious part of this change are in separate commits:
* Testing write-without-read file permission behavior was removed from the `p2_file_stream_not_permitted` test program, and its test harness.
* The `p{1,2,3}_file_rename_across_perms` tests were changed to just test that a rename was not permitted, because the weird corner case from the previous model of creating an intermediate hardlink is now disallowed.

### CODE REVIEW

File: crates/wasi/src/filesystem.rs [hunk 1]
```diff
@@ -1072,7 +1059,7 @@ impl Dir {
     }
 
     pub(crate) async fn readlink_at(&self, path: String) -> Result<String, ErrorCode> {
-        if !self.perms.contains(DirPerms::READ) {
+        if self.perms.write_not_permitted() {
             return Err(ErrorCode::NotPermitted);
         }
```
Comment (alexcrichton) [inline crates/wasi/src/filesystem.rs]: I think this check can probably be removed entirely

**Review by alexcrichton — APPROVED** [review]

File: crates/wasi/src/filesystem.rs (hunk 1 above)
Comment (pchickey) [reply crates/wasi/src/filesystem.rs]: Thanks for catching that, fixed

File: crates/wasi/src/filesystem.rs [hunk 2]
```diff
@@ -76,36 +76,44 @@ pub trait WasiFilesystemView: Send {
...
+    ReadWrite,
 }
 
-bitflags::bitflags! {
-    #[derive(Copy, Clone, Debug, PartialEq, Eq)]
-    pub struct OpenMode: usize {
-        const READ = 0b1;
-        const WRITE = 0b10;
+impl FsPerms {
+    /// Convenience function, t
```
Comment (dicej) [inline crates/wasi/src/filesystem.rs]: Looks like this comment was interrupted.

**Review by dicej — APPROVED** [review]

**Review by rvolosatovs — APPROVED** [review]
